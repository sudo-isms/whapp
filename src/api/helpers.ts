// normalize response to a collection + array of Ids
// saves some array.prototype.find()'s and
// makes working with related data easier
export function normalizeResponse(data = []) {
  const entities = { byId: {}, allIds: [] };
  data.forEach((entity) => {
    entities.byId[entity.id] = entity;
    entities.allIds.push(entity.id);
  });
  return entities;
}

type DateSortable = { createdAt: string };

export function sortByCreatedAt(a: DateSortable, b: DateSortable) {
  return Date.parse(b.createdAt) - Date.parse(a.createdAt);
}
