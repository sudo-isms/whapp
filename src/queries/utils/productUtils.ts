// amends the amountInStock property to the products
// after fetching data. This could very well be broken out into
// a computed value, but this way is quite convenient if we often
// want to display just the product listing, sans articles
export function amendProductsInStock(products, articles) {
  if (!products || !articles) return false;
  products.allIds.forEach((productId) => {
    const product = products.byId[productId];
    if (!product) {
      return false;
    }

    let amountInStock = null;

    // Articles related to a product === part
    product.articles.some((part) => {
      const availableArticles = articles.byId[part.id].amountInStock;
      // if the part is not in stock, we can't assemble the product
      if (availableArticles === 0) {
        amountInStock = 0;
        return true;
      }
      const partCount = Math.floor(availableArticles / part.amountRequired);
      if (amountInStock === null || partCount < amountInStock) {
        amountInStock = partCount;
      }
      return false;
    });

    product.amountInStock = amountInStock;
  });
  return products;
}
