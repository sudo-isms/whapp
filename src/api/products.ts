import { normalizeResponse, sortByCreatedAt } from './helpers';

const apiUrl = process.env.REACT_APP_API_URL || 'http://localhost:7000';

const defaultHeaders = {
  'Content-Type': 'application/json',
};

// PRODUCTS

export function getProducts() {
  return fetch(apiUrl + '/products')
    .then((response) => {
      if (!response.ok) {
        return Promise.reject(response);
      }
      return response.json();
    })
    .then(normalizeResponse);
}

// ARTICLES

export async function getArticles() {
  return fetch(apiUrl + '/articles')
    .then((response) => {
      if (!response.ok) {
        return Promise.reject(response);
      }
      return response.json();
    })
    .then(normalizeResponse);
}

// SALES

type Sale = { productId: string; amountSold: number };

export async function getSales() {
  return fetch(apiUrl + '/sales').then((response) => {
    if (!response.ok) {
      return Promise.reject(response);
    }
    // sort by createdAt in descending order
    return response.json().then((sales) => {
      if (Array.isArray(sales)) {
        return sales.sort(sortByCreatedAt);
      }
      return sales;
    });
  });
}

export async function postSale({ productId, amountSold }: Sale) {
  return fetch(apiUrl + '/sales', {
    method: 'POST',
    headers: defaultHeaders,
    body: JSON.stringify({ productId, amountSold }),
  }).then((response) => {
    if (!response.ok) {
      return Promise.reject(response);
    }
    return response.json();
  });
}

export async function deleteSale(id: string) {
  return fetch(`${apiUrl}/sales/${id}`, {
    method: 'DELETE',
  }).then((response) => {
    if (!response.ok) {
      return Promise.reject(response);
    }
    return Promise.resolve(response);
  });
}

// ARTICLES

export function patchArticles({ articles }) {
  return fetch(apiUrl + '/articles', {
    method: 'PATCH',
    headers: defaultHeaders,
    body: JSON.stringify(articles),
  }).then((response) => {
    if (!response.ok) {
      return Promise.reject(response);
    }
    return response.json();
  });
}

// > NOTE: You can only patch `name` and `amountInStock`.
// > You can also use `amountToSubtract` to patch the _amount in stock_.
// > The value provided will be subtracted from`amountInStock`.
//
// Harder to achieve resilient transaction if using `amountToSubtract`,
// given that the api is not stable. Makes it harder to replay a failed mutation.
// Would be preferrable if the backend handled the changes to article stock,
// when receiving a sale, but then it would also make for a less fun assignment.
export function updateArticleStock({ productId, amountSold, queryCache }) {
  const products = queryCache.getQueryData('products');
  const articles = queryCache.getQueryData('articles');

  const product = products.byId[productId];
  let articlesToPatch = [];

  try {
    articlesToPatch = product.articles.map((item) => {
      const { amountInStock: currentAmountInStock, name } =
        articles.byId[item.id] || {};

      const amountInStock =
        currentAmountInStock - item.amountRequired * amountSold;

      if (amountInStock < 0) {
        throw new Error(
          `${name} (${item.id}) will have a negative amount in Stock: ${amountInStock}`
        );
      }

      return {
        id: item.id,
        amountInStock,
      };
    });
  } catch (err) {
    return Promise.reject(err);
  }

  return fetch(apiUrl + '/articles', {
    method: 'PATCH',
    headers: defaultHeaders,
    body: JSON.stringify(articlesToPatch),
  }).then((response) => {
    if (!response.ok) {
      return Promise.reject(response);
    }
    return Promise.resolve(response);
  });
}
