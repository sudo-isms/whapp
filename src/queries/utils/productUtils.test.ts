import { amendProductsInStock } from './productUtils';
import { normalizeResponse } from '../../api/helpers';

import rawArticles from './__mocks__/articles.json';
import rawProducts from './__mocks__/products.json';

const articles = normalizeResponse(rawArticles);
const products = normalizeResponse(rawProducts);

describe('queries: amendProductsInStock', () => {
  test('will calculate and amend amountInStock to a product', () => {
    amendProductsInStock(products, articles);

    expect(products).toMatchSnapshot();
  });
});
