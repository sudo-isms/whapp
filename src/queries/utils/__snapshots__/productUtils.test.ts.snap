// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries: amendProductsInStock will calculate and amend amountInStock to a product 1`] = `
Object {
  "allIds": Array [
    "a269a247-0d38-4b47-9630-79c9ae545b68",
    "6fed6191-ee01-4563-a33d-5010abe0db36",
  ],
  "byId": Object {
    "6fed6191-ee01-4563-a33d-5010abe0db36": Object {
      "amountInStock": 1,
      "articles": Array [
        Object {
          "amountRequired": 4,
          "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
        },
        Object {
          "amountRequired": 1,
          "id": "6892b98b-9b87-4520-9a9e-7528f1d78cb4",
        },
        Object {
          "amountRequired": 8,
          "id": "addc65a8-c759-41d8-a18a-89fe446ad484",
        },
      ],
      "id": "6fed6191-ee01-4563-a33d-5010abe0db36",
      "name": "Dining Table",
    },
    "a269a247-0d38-4b47-9630-79c9ae545b68": Object {
      "amountInStock": 2,
      "articles": Array [
        Object {
          "amountRequired": 4,
          "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
        },
        Object {
          "amountRequired": 1,
          "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
        },
        Object {
          "amountRequired": 8,
          "id": "addc65a8-c759-41d8-a18a-89fe446ad484",
        },
      ],
      "id": "a269a247-0d38-4b47-9630-79c9ae545b68",
      "name": "Dining Chair",
    },
  },
}
`;
